### DISCUSSION TITLE
BUG : Consultation de la vue logical_infrastructure

### DISCUSSION THREAD

**Discussion by SteevensICam** [dbarzin/mercator#511] [Q&A]
Bonjour,

Nous rencontrons un problème sur la consultation de la vue logique infra, et la création d'un rapport de la vue infra logique.

![image](https://github.com/dbarzin/mercator/assets/142876160/210c6808-eec6-4e0a-b07b-2fced0263c03)

![image](https://github.com/dbarzin/mercator/assets/142876160/7720c32c-6492-4eb0-bab3-58fc6158b054)

Nous avons ceci comme erreur : 

![image](https://github.com/dbarzin/mercator/assets/142876160/65a8f128-64ed-4063-8f0b-16e218950530)

Nous sommes sur la dernière version, Maj réalisée le 25/09/2023 à 9h20.

Pourriez-vous y remédier ? 
Je vous en remercie.

**Comment by SteevensICam** [comment]
Bonjour, en complément, nous rencontrons également un problème à la création dans : Vue des infrastructures logique / Entité extérieure connectée.

![image](https://github.com/dbarzin/mercator/assets/142876160/09f027de-f8aa-4551-8783-768f4b55333d)

Cordialement.

**Reply by SteevensICam** [reply]
pour reprendre ce sujet là

dans un docker tout beau tout neuf, lorsque nous inspectons le schéma de la table _external_connected_entities_
voici le résultat:
```
sqlite> .schema external_connected_entities
CREATE TABLE external_connected_entities (id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, name VARCHAR(255) NOT NULL COLLATE "BINARY", contacts VARCHAR(255) DEFAULT NULL COLLATE "BINARY", created_at DATETIME DEFAULT NULL, updated_at DATETIME DEFAULT NULL, deleted_at DATETIME DEFAULT NULL);
CREATE UNIQUE INDEX external_connected_entities_name_unique ON external_connected_entities (name, deleted_at);
CREATE INDEX "entity_id_fk_1295034" on "external_connected_entities" ("entity_id");
CREATE INDEX "network_id_fk_8596554" on "external_connected_entities" ("network_id");
```
ce qui est cohérent avec la migration 2021_05_08_191249_create_external_connected_entities_table.php

```
 public function up()
    {
        Schema::create('external_connected_entities', function (Blueprint $table) {
            $table->increments('id');
            $table->string('name');
            $table->string('responsible_sec')->nullable();
            $table->string('contacts')->nullable();
            $table->timestamps();
            $table->softDeletes();
            $table->unique(['name', 'deleted_at'], 'external_connected_entities_name_unique');
        });
    }
```
et avec la documentation aussi

![image](https://github.com/dbarzin/mercator/assets/142876160/e6389e65-6241-4d96-82ae-74d0c44ff9fa)


et comme si la migration _2022_08_11_165441_add_vpn_fields.php_ n'était pas passée, logiquement on ne peut pas ajouter 

> Créer Entité extérieure connectée  

`SQLSTATE[HY000]: General error: 1 table external_connected_entities has no column named type (Connection: sqlite, SQL: insert into "external_connected_entities" ("name", "type", "entity_id", "contacts", "description", "network_id", "src", "dest", "updated_at", "created_at") values (aaa, aa, 1, dfgdfg, aeae, ?, 192.168.0.1, 192.168.03.254, 2023-10-31 14:04:54, 2023-10-31 14:04:54))`

le php artisan migrate:status est ok pour toutes.

Comme puis-je aider ? :)

**Reply by dbarzin** [reply]
Je pense que ce problème est corrigé dans le reste du fil de cette discussion.

**Reply by SteevensICam** [reply]
Hélas non
dans un environnement docker
sur installation vierge, le problème persiste, je nous le confirme à l'instant !

Je n'ai pas réalisé de test d'installation hors docker pour faire un test croisé.

**Reply by SteevensICam** [reply]
mon constat actuel :

- j'ai supprimé la migration 155 2022_08_11_165441_add_vpn_fields dans la table migrations 
- vérifié avec php artisan migrate:status 
![image](https://github.com/dbarzin/mercator/assets/142876160/9911ba29-8810-4b0a-ae36-cb1fc81bcfc8)
- rejoué la migration et arrive sur cette erreur.
![image](https://github.com/dbarzin/mercator/assets/142876160/14403f00-d843-4165-aabd-aaa341d673c6)

je ne sais pas si cela peut vous aider.

**Reply by SteevensICam** [reply]
https://github.com/dbarzin/mercator/blob/master/database/migrations/2022_08_11_165441_add_vpn_fields.php

il me semble qu'il pourrait y avoir une inversion dans ce fichier:
ligne 20 dans la methode up on à 
`            $table->dropColumn("responsible_sec");`

et dans la méthode down on à :
`            $table->string("responsible_sec")->nullable()->after("name");`

ce qui correspondrait bien à l'erreur que je constate dans mon message précédent.

**Reply by dbarzin** [reply]
La migration est correcte. Voici la structure de la table actuelle en production

```SQL
mysql> describe external_connected_entities;
+-------------+--------------+------+-----+---------+----------------+
| Field       | Type         | Null | Key | Default | Extra          |
+-------------+--------------+------+-----+---------+----------------+
| id          | int unsigned | NO   | PRI | NULL    | auto_increment |
| name        | varchar(255) | NO   | MUL | NULL    |                |
| description | longtext     | YES  |     | NULL    |                |
| type        | varchar(255) | YES  |     | NULL    |                |
| entity_id   | int unsigned | YES  | MUL | NULL    |                |
| network_id  | int unsigned | YES  | MUL | NULL    |                |
| src         | varchar(255) | YES  |     | NULL    |                |
| dest        | varchar(255) | YES  |     | NULL    |                |
| contacts    | varchar(255) | YES  |     | NULL    |                |
| created_at  | timestamp    | YES  |     | NULL    |                |
| updated_at  | timestamp    | YES  |     | NULL    |                |
| deleted_at  | timestamp    | YES  |     | NULL    |                |
+-------------+--------------+------+-----+---------+----------------+
12 rows in set (0,01 sec)
```

Le Docker a été testé et fonctionne.

**Reply by dbarzin** [reply]
Je n'arrive pas à reproduire. Peux-tu m'aider à reproduire le problème ?

**Reply by SteevensICam** [reply]
Step to reproduce :

*  debian 11.8 fresh install
* commandes présente dans la doc
```
docker pull ghcr.io/dbarzin/mercator:latest
docker run -d --name mercator  -p "127.0.0.1:8000":80 ghcr.io/dbarzin/mercator:latest
docker exec -it -u root mercator bash
```

puis 

`bash-5.1# sqlite3 sql/db.sqlite`

et j'affiche le schéma de notre table:

`sqlite> .schema external_connected_entities
CREATE TABLE external_connected_entities (id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, name VARCHAR(255) NOT NULL COLLATE "BINARY", contacts VARCHAR(255) DEFAULT NULL COLLATE "BINARY", created_at DATETIME DEFAULT NULL, updated_at DATETIME DEFAULT NULL, deleted_at DATETIME DEFAULT NULL);
CREATE UNIQUE INDEX external_connected_entities_name_unique ON external_connected_entities (name, deleted_at);
CREATE INDEX "entity_id_fk_1295034" on "external_connected_entities" ("entity_id");
CREATE INDEX "network_id_fk_8596554" on "external_connected_entities" ("network_id");
`

**Comment by dbarzin** [comment]
Avez-vous bien migré la base de données ?

    php artisan migrate


Y a-t-il des erreurs lors de la migration ?

**Comment by SteevensICam** [comment]
bonjour, non nous n'avons pas de message d'erreur concernant cette commande.

mais en creusant, nous pensons avoir un problème de gestion de notre mise à jour docker.

explication :)

notre db.sqlite est
* stockée dans le docker.
* sauvegardée régulièrement 
* remise dans le docker suite à une mise à jour.

=> à priori lorsque l'on remet la sauvegarde, c'est logique, on écrase le db.sqlite local, et donc on conserve l'ancien schéma !

Question : peut-on rejouer les migrate précédents ?
sinon, alors nous devrons utiliser l'externalisation selon la doc : 
-v /var/www/mercator/db.sqlite:/var/www/mercator/db.sqlite 

Qu'en pensez-vous ?
Merci par avance.

**Comment by dbarzin** [comment]
Vous devez suivre la procédure de mise à jour décrite ici : https://github.com/dbarzin/mercator/blob/master/INSTALL.fr.md#mise-%C3%A0-jour. Pour mettre à jour la base de données, c'est la commande :

    php artisan migrate

qui effectue les changements dans la base.

**Comment by SteevensICam** [comment]
ok fait, la dernière ne passe pas.
![image](https://github.com/dbarzin/mercator/assets/142876160/5958839b-76b4-4c3e-912f-f2dbcbb71c4d)

**Comment by dbarzin** [comment]
La commande de création de la table "clusters" n'est sans doute pas la vrais raison.
Est-ce la première fois que cette migration a échouée ou a -t-elle déjà échouée auparavant ?
Est-il possible de noter le premier échec de migration ?
Autre idée, le mot "clusters" est-il réservé dans SQLite ?

**Reply by SteevensICam** [reply]
Alors là nous utilisons docker
de la manière suivante :

docker pull ghcr.io/dbarzin/mercator:latest
docker stop mercator
docker rm mercator

puis
docker run  -d --name mercator  -p "127.0.0.1:8000":80 ghcr.io/dbarzin/mercator:latest

Nous n'avons pas essayé de restaurer notre base.
c'est, je crois, une installation from scratch.

**Reply by dbarzin** [reply]
Merci, on va vérifier cela.

**Reply by SteevensICam** [reply]
avec plaisir :)
On va y arriver, mais là, je vous avoue sécher un peu sur cette histoire

**Reply by dbarzin** [reply]
OK, corrigé !
